### PR TITLE
ARROW-15515: [C++] Update ExecPlan example code and documentation with new options

### PR DESCRIPTION
**Pull request by vibhatha**
This PR includes a documentation update for streaming execution engine and a `table_sink` example.

### CODE REVIEW

**Comment by vibhatha** [comment]
cc @westonpace included the `table_sink` option. I added a separate example for this and updated the docs. 

Do you think we should replace the collector function `ExecutePlanAndCollectAsTable` with a `table_sink` ?

File: cpp/examples/arrow/execution_plan_documentation_examples.cc [hunk 1]
```diff
@@ -851,6 +851,50 @@ arrow::Status SourceUnionSinkExample(cp::ExecContext& exec_context) {
...
+
+  auto finish = source->finished();
+
+  RETURN_NOT_OK(finish.status());
+
+  std::cout << "Results : " << output_table->ToString() << std::endl;
+
+  // plan mark finished
+  auto future = plan->finished();
+  return future.status();
```
Comment (westonpace) [inline cpp/examples/arrow/execution_plan_documentation_examples.cc:894]:
```suggestion
  // Wait for the plan to finish
  auto finished = plan->finished();
  RETURN_NOT_OK(finished.status());

  std::cout << "Results : " << output_table->ToString() << std::endl;
  return Status::OK();
```
I don't think it's a good idea to do `source->finished()` as I don't think it's a good idea to wait on individual nodes.  User's should probably only ever wait on the plan.

File: docs/source/cpp/streaming_execution.rst [hunk 2]
```diff
@@ -647,6 +649,25 @@ SelectK example:
...
 .. _stream_execution_scan_docs:
 
+``table_sink``
+----------------
+
+.. _stream_execution_table_sink_docs:
+
+Considering the variety of sink nodes provided in the streaming execution engine, the ``table_sink`` node 
+provides the ability to take the output as a table. It is much easier to use 
+:class:`arrow::compute::TableSinkNodeOptions`.
```
Comment (westonpace) [inline docs/source/cpp/streaming_execution.rst:659]:
```suggestion
The ``table_sink`` node provides the ability to take the output as an in-memory table. 
This is much simpler to use than the other sink nodes provided by the streaming execution engine
but it only makes sense when the output fits comfortably in memory.  The node is created using :class:`arrow::compute::TableSinkNodeOptions`.
```

File: docs/source/cpp/streaming_execution.rst [hunk 3]
```diff
@@ -647,6 +649,25 @@ SelectK example:
...
+The output data can be obtained as a ``std::shared_ptr<arrow::Table>`` along with the output ``schema``.
```
Comment (westonpace) [inline docs/source/cpp/streaming_execution.rst:660]: Technically it was an input schema I think (and it's going away now).  A `Table` already has a schema associated with it so there is no need to specify the output schema separately.

**Review by westonpace — COMMENTED** [review]
Looks good.  I added a few wording suggestions.  This will conflict with #12721 so one of the two will need to rebase (which gets merged second).

**Comment by vibhatha** [comment]
@westonpace Thank you for the review. I will address these issues.

File: docs/source/cpp/streaming_execution.rst [hunk 4]
```diff
@@ -647,6 +649,25 @@ SelectK example:
 
 .. _stream_execution_scan_docs:
 
+``table_sink``
+----------------
+
+.. _stream_execution_table_sink_docs:
+
+The ``table_sink`` node provides the ability to take the output as an in-memory table. 
+This is much simpler to use than the other sink nodes provided by the streaming execution engine
```
Comment (westonpace) [inline docs/source/cpp/streaming_execution.rst:658]:
```suggestion
The ``table_sink`` node provides the ability to receive the output as an in-memory table. 
This is simpler to use than the other sink nodes provided by the streaming execution engine
```
Optional, but this sounds slightly better.

File: cpp/src/arrow/compute/exec/options.h [hunk 5]
```diff
@@ -176,6 +176,20 @@ class ARROW_EXPORT OrderBySinkNodeOptions : public SinkNodeOptions {
   SortOptions sort_options;
 };
 
+/// \brief Adapt an Table as a sink node
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/options.h:179]:
```suggestion
/// \brief Adapt a Table as a sink node
```

File: cpp/src/arrow/compute/exec/options.h [hunk 6]
```diff
@@ -176,6 +176,20 @@ class ARROW_EXPORT OrderBySinkNodeOptions : public SinkNodeOptions {
...
+///
+/// obtains the output of a execution plan to
+/// a table pointer.
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/options.h:182]:
```suggestion
/// obtain the output of an execution plan to
/// a table.
```

**Review by westonpace — APPROVED** [review]
Thanks!  Just a few minor grammatical suggestions.

**Pull request closed without merging**